91/migrate tool

helenwalsh opened the pull request:
V9.1.1 adding the migrate code which is compatible with PyU4V9.1

Commit:
test

Commit:
Adding migrate tool to PyU4V

This is already in 3.1.x but needed migrating to PyU4V 9.1.
Uses non deprecated methods
Abides by PyU4V 9.1 coding conventions.

Commit:
Adding migrate tool to PyU4V

This is already in 3.1.x but needed migrating to PyU4V 9.1.
Uses non deprecated methods
Abides by PyU4V 9.1 coding conventions.
Update of documentation to add new tool section.
Version is 9.1.1.0

MichaelMcAleer commented on docs/build/html/_sources/contribute.rst.txt:
```diff
@@ -35,7 +35,7 @@ forward on all code in PyU4V.
 2. Use ``.format()`` for string concatenation.
 
-.. code-block::
+.. code-block:: text
```
Code type should be python here

helenwalsh replied:
It complained about it when I put in 'Python'.  Let me try it again

MichaelMcAleer commented on ChangeLog:
```diff
@@ -1,328 +1,332 @@
-  console driven commands from your own scripts and not from PyU4V
+PyU4V Change Log
+================
+
```
You need to add change log details about the changes in this pull request

helenwalsh replied:
I put it down the end.  I will bring it up

MichaelMcAleer commented on docs/source/contribute.rst:
```diff
@@ -35,7 +35,7 @@ forward on all code in PyU4V.
 2. Use ``.format()`` for string concatenation.
 
-.. code-block::
+.. code-block:: text
```
Code block should be Python

helenwalsh replied:
Again it complained on the make html.  I will take another look

MichaelMcAleer commented on docs/source/tools.rst:
```diff
@@ -0,0 +1,73 @@
+
+Tools Guide
+=================
```
Title underline === should be same length as the title, I got warnings about this in my own environment when I was working on the 91 docs

helenwalsh replied:
I did not get any warnings, but yes needs to be changed

odonos12 replied:
Make underline === same length as Tools Guide.

MichaelMcAleer commented on docs/source/tools.rst:
```diff
@@ -0,0 +1,73 @@
+release.
+
+.. important::
+   - Running this script is not necessary unless you intend 'Live Migrating' from one compute node to another.
```
These lines need wrapped at 79 chars where possible, there are some tables elsewhere that are impossible to get into that size but they are the only exception in the .rst files.

helenwalsh replied:
sure

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+        """
+        self.conn = conn
+
+    def check_input(self, txt_str, option):
```
Method is static, will we keep conformity with the rest of PyU4V and mark static methods as static? There are a number of them in this file but ill keep it to this one comment about the collection of them as a whole

odonos12 replied:
@staticmethod decorator needed

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+
+        return component_dict
+
+    def check_mv_for_migration(self, mv_name, revert=False):
```
revert isn't in docstring

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+        """Parse the storage group string.
+
+        :param sg_name: the storage group name
+        :param test: is it a test case
```
This isnt an input param

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+        return self.get_object_components_and_correct_host(regex_str, sg_name)
+
+    def get_element_dict_revert(self, component_dict, sg_name, cd_str, re_str,
+                              pg_name, host_name):
```
second row not aligned

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+        return element_dict
+
+    def compile_new_element_names(
+            self, mv_name, pg_name, host_name, sg_name, revert=False):
```
revert isnt in docstring

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+        :param pg_name: port group name
+        :param host_name: host name
+        :param sg_name: storage group name
+        :param test: is it a test case
```
test isnt an input param

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+                        self.conn.provisioning.add_child_sg_to_parent_sg(
+                            element_dict['new_sg_name'],
+                            element_dict['new_sg_parent_name']))
+                    self.get_storage_group(message, DEBUG)
```
Unexpected argument, only one input defined in get_storage_group

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+        return masking_view
+
+    def get_or_create_masking_view(
+            self, element_dict, portgroup, host, revert=False):
```
revert isn't in docstring

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+        :param element_dict: element dictionary
+        :param portgroup: port group name
+        :param host: host name
+        :param test: is it a test case
```
test isn't an input param

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+        :param storage_group_name: the storage group name
+        :returns: List, Boolean
+        """
+        create_vol = False
```
create_vol here is redundant, it is set in both if and else code blocks

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+
+        return volume_list, create_vol
+
+    def choose_sg(
```
storage_group

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+        return volume_list, create_vol
+
+    def choose_sg(
+            self, masking_view, child_sg_list, portgroup, initiatorgroup,
```
initiator_group, port_group etc

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+            try:
+                if revert:
+                    return True if self.get_mv_component_dict(
+                        masking_view, sys.argv[1]) else False
```
sys.argv[1] may need cast as a boolean here, get_mv_component_dict() expects boolean value for revert input param

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+                object_dict['host'] = object_dict['host'][:-1]
+        return object_dict
+
+    def get_mv_component_dict(self, mv_name, revert=False):
```
PyU4V standards: masking_view

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+                self.smart_print(print_str, WARNING, mv_name)
+                return False
+
+    def get_sg_component_dict(self, sg_name):
```
storage_group

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+                element_dict['new_mv_name'])
+        return new_masking_view_details
+
+    def move_vols_from_source_to_target(
```
vols is fine here but rest of methods use volumes if you want to keep it consistent

helenwalsh replied:
sure

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,866 @@
+            self.smart_print('Problem with input stream', ERROR)
+        return txt_out
+
+    def get_sg_qos_details(self, sg_details):
```
storage_group

MichaelMcAleer commented on PyU4V/tools/openstack/migrate.py:
```diff
@@ -0,0 +1,149 @@
+                            element_dict, mv_components['portgroup'],
+                            mv_components['initiatorgroup']))
+            # Get the volumes in the storage group
+            if 'storagegroup' in mv_components:
```
mv_components may be referenced before assignment, its in the first if block but the else block has an if in there also that may not be met

MichaelMcAleer commented on PyU4V/tools/openstack/migrate.py:
```diff
@@ -0,0 +1,149 @@
+            if 'storagegroup' in mv_components:
+                # Check the qos setting of source and target storage group
+                utils.set_qos(mv_components['storagegroup'],
+                              element_dict['new_sg_name'])
```
same as element_dict, is may also be referenced before assignment for same reason as mv_components

MichaelMcAleer reviewed (changes requested):
Few extra things required that I couldnt get into the review comments:
- Version needs updated in:
utils/constants.py - line 33 PYU4V_VERSION
docs/source/index.rst - in the table
docs/source/installation.rst - in the table

I think that tools may need added to setup.py, it has utils listed in there explicitly:
`packages=['PyU4V', 'PyU4V.utils'],`

When you click on tools guide in docs included in this pull request it returns a 404 not found error.

Commit:
Adding migrate tool to PyU4V

This is already in 3.1.x but needed migrating to PyU4V 9.1.
Uses non deprecated methods
Abides by PyU4V 9.1 coding conventions. IMPROVED
Update of documentation to add new tool section.
Version is 9.1.1.0
Change to tox.ini as it was not including openstack for pep8
ChangeLog entry moved.

Commit:
Adding migrate tool to PyU4V

This is already in 3.1.x but needed migrating to PyU4V 9.1.
Uses non deprecated methods
Abides by PyU4V 9.1 coding conventions. IMPROVED
Update of documentation to add new tool section.
Version is 9.1.1.0
Change to tox.ini as it was not including openstack for pep8
ChangeLog entry moved.
Html updated to add type to method arguments and return values

Commit:
Adding migrate tool to PyU4V

This is already in 3.1.x but needed migrating to PyU4V 9.1.
Uses non deprecated methods
Abides by PyU4V 9.1 coding conventions. IMPROVED
Update of documentation to add new tool section.
Version is 9.1.1.0
Change to tox.ini as it was not including openstack for pep8
ChangeLog entry moved.
Html updated to add type to method arguments and return values

odonos12 commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,911 @@
+        """Check if the masking view can be migrated.
+
+        :param masking_view_name: the masking view name -- str
+        :param test: is it a test case -- boolean
```
:param test: -> :param revert:

MichaelMcAleer replied:
Shouldn't pep8 be catching these also?

odonos12 commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,911 @@
+        """Parse the storage group string.
+
+        :param storage_group_name: the storage group name -- str
+        :param test: is it a test case -- boolean
```
Remove :param test:

odonos12 commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,911 @@
+        :param port_group_name: port group name -- str
+        :param host_name: host name -- str
+        :param storage_group_name: storage group name -- str
+        :param test: is it a test case -- boolean
```
:param test: -> :param revert:

odonos12 commented on docs/build/html/_static/documentation_options.js:
```diff
@@ -1,10 +1,10 @@
     HAS_SOURCE: true,
     SOURCELINK_SUFFIX: '.txt',
-    NAVIGATION_WITH_KEYS: false
+    NAVIGATION_WITH_KEYS: false,
```
Should the trailing "," be here?

MichaelMcAleer replied:
These are autogenerated files during sphinx build so I wouldn't worry too much about problems in the docs folder that are outside of the docs/source directory

odonos12 commented on PyU4V/tests/unit_tests/test_migrate.py:
```diff
@@ -0,0 +1,960 @@
+        self.utils = migrate_utils.MigrateUtils(conn)
+
+    def test_get_masking_view_component_dict_old(self):
+        """Test for get_masking_view_component_dict."""
```
Same docstring used across multiple tests in this file, is it possible to append a descriptive unique identifier to the end for these comments so that each test has a unique docstring?

odonos12 reviewed (changes requested):
Minor docstring issues.

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,911 @@
+
+from PyU4V.utils import exception
+
+import prettytable
```
Pretty Table will need listed in setup.py as a package dependency and also in our docs as a requirement because it isn't  a native Python library

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,911 @@
+        return masking_view
+
+    def get_or_create_masking_view(
+            self, element_dict, port_group, host, revert=False):
```
revert needs added to docstring and test removed

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,911 @@
+        :param storage_group_name: the storage group name -- str
+        :returns: volume list -- list, create volume -- boolean
+        """
+        create_volume_flag = False
```
This value isn't needed, create_volume_flag is set below in if/else blocks

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,911 @@
+            try:
+                if revert:
+                    return True if self.get_masking_view_component_dict(
+                        masking_view_name, sys.argv[1]) else False
```
sys.argv[1] needs converted to a boolean dependent on user input string, expected bool got str

MichaelMcAleer commented on PyU4V/tools/openstack/migrate.py:
```diff
@@ -0,0 +1,192 @@
+                is_revert))
+        # Check if masking view exists and if it does validate it
+        if element_dict:
+                utils.get_or_create_masking_view(
```
This utils block is over indented, it needs to come back a single indent

MichaelMcAleer commented on PyU4V/tools/openstack/migrate.py:
```diff
@@ -0,0 +1,192 @@
+            element_dict['new_sg_name'])
+        utils.print_pretty_table(new_storage_group)
+
+for masking_view in masking_view_list:
```
Two lines between methods outside of classes, getting a pep8 warning in PyCharm but not in the pep8 tox tun

MichaelMcAleer commented on setup.py:
```diff
@@ -16,7 +16,7 @@
 
 setup(name='PyU4V',
-      version='9.1.0.0',
+      version='9.1.1.0',
```
Does PyU4V.tools need to be added to packages in the same manner as PyU4V.utils?

MichaelMcAleer commented on docs/source/tools.rst:
```diff
@@ -0,0 +1,79 @@
+=================
+
+First, make sure that PyU4V is `installed <http://pyu4v.readthedocs.io/en/latest/installation.html>`_.
+Then visit the `quick_start <http://pyu4v.readthedocs.io/en/latest/quick_start.html>`_ section to make
```
Lines run over 79 chars, you can shorten links by adding them to the bottom of the file in the same way as the other files

MichaelMcAleer replied:
This still needs addressed, I think specifying URL links at the bottom is a better approach than in-line, there is already a URLs section at the bottom of the tools page.

MichaelMcAleer commented on docs/source/contribute.rst:
```diff
@@ -35,7 +35,7 @@ forward on all code in PyU4V.
 2. Use ``.format()`` for string concatenation.
 
-.. code-block::
+.. code-block:: Python
```
I have no idea how they got to be the incorrect single quote but:
`variable1=’Hello’, variable2=’PyU4V’)`
Should be...
`variable1='Hello', variable2='PyU4V')`

MichaelMcAleer replied:
This still needs addressed, the error is a layover from our work on the docs that was never spotted, if we are changing contribute.rst to specify the block as Python we should correct the code at the same time

MichaelMcAleer reviewed (changes requested):
A lot of the comments were addressed in recent commits but there are still some unresolved one in there and a few new ones.

Commit:
Adding migrate tool to PyU4V

This is already in 3.1.x but needed migrating to PyU4V 9.1.
Uses non deprecated methods
Abides by PyU4V 9.1 coding conventions. IMPROVED
Update of documentation to add new tool section.
Version is 9.1.1.0
Change to tox.ini as it was not including openstack for pep8
ChangeLog entry moved.
Html updated to add type to method arguments and return values
setup.py changed to add missing packages

Commit:
Adding migrate tool to PyU4V

This is already in 3.1.x but needed migrating to PyU4V 9.1.
Uses non deprecated methods
Abides by PyU4V 9.1 coding conventions. IMPROVED
Update of documentation to add new tool section.
Version is 9.1.1.0
Change to tox.ini as it was not including openstack for pep8
ChangeLog entry moved.
Html updated to add type to method arguments and return values

helenwalsh committed:
Merge branch 'master' into 91/migrate_tool

MichaelMcAleer commented on PyU4V/tools/openstack/migrate_utils.py:
```diff
@@ -0,0 +1,910 @@
+        else:
+            return False
+
+    def print_to_log(self, print_str, level=DEBUG):
```
For succinctness there was only one comment added about methods not being marked as @static when they should be, there is a lot more instances throughout this file, print_to_log is another example, print_pretty_table, verify_protocol etc

MichaelMcAleer replied:
Pycharm should highlight them all infile and in the scroll bar as warnings

MichaelMcAleer commented on ChangeLog:
```diff
@@ -1,336 +1,337 @@
+================
+
+Version 9.1.1.0 - released 12/12/19
+==================================
```
Title underscore is short a =

MichaelMcAleer commented on setup.py:
```diff
@@ -23,8 +23,8 @@
-      packages=['PyU4V', 'PyU4V.utils'],
-      install_requires=['requests', 'six', 'urllib3'],
+      packages=['PyU4V', 'PyU4V.utils', 'PyU4V.tools.openstack'],
+      install_requires=['requests', 'six', 'urllib3', 'prettytable', 'future'],
```
`future` is a Python native library it doesn't need listed as a requirement

helenwalsh replied:
I had to add it for python2.7 but I guess not required if we no longer support 2.7

MichaelMcAleer commented on setup.py:
```diff
@@ -23,8 +23,8 @@
+      install_requires=['requests', 'six', 'urllib3', 'prettytable', 'future'],
       include_package_data=True,
       classifiers=[
           'Development Status :: 2 - Pre-Alpha',
```
Here is something else we overlooked, our development status should be:

`Development Status :: 5 - Production/Stable`

https://pypi.org/classifiers/

helenwalsh replied:
Sure, I'll change it.

MichaelMcAleer commented on setup.py:
```diff
@@ -23,8 +23,8 @@
       license='Apache 2.0',
-      packages=['PyU4V', 'PyU4V.utils'],
-      install_requires=['requests', 'six', 'urllib3'],
+      packages=['PyU4V', 'PyU4V.utils', 'PyU4V.tools.openstack'],
```
Still scratching head over this one and how to best approach it, it came from this customer issue: https://github.com/MichaelMcAleer/PyU4V/issues/11 

I'm not sold on the prospect of having to list out all packagaes, sub-packaged, sub-subpackages etc.  Had a quick look online, we may be able to switch to `packages=setuptools.find_packages()` as directed in https://packaging.python.org/tutorials/packaging-projects/ under `'Creating setup.py'` and again here https://packaging.python.org/guides/distributing-packages-using-setuptools/#packages. 

Taking this approach and changing may save us from having to hard-code any packages, sub-packages, sub-subpackages etc moving forward.  One to discuss next time we meet.

helenwalsh replied:
Absolutely.  I can give your suggestion a quick test.

Commit:
Adding migrate tool to PyU4V

This is already in 3.1.x but needed migrating to PyU4V 9.1.
Uses non deprecated methods
Abides by PyU4V 9.1 coding conventions. IMPROVED
Update of documentation to add new tool section.
Version is 9.1.1.0
Change to tox.ini as it was not including openstack for pep8
ChangeLog entry moved.
Html updated to add type to method arguments and return values

MichaelMcAleer reviewed (approved):
All this looks fine now, I think the minor version should be 9.1.2.0 because we are on 9.1.1.0 on master but this is something we should iron out in our next team meeting
